ydb/core/tx/schemeshard/schemeshard__operation_replace_external_table.cpp

```diff
@@ -0,0 +1,421 @@
+        context.SS->PersistTxState(db, OperationId);
+    }
+
+public:
+    using TSubOperation::TSubOperation;
+
+    THolder<TProposeResponse> Propose(const TString& owner, TOperationContext& context) override {
+        Y_UNUSED(owner);
+        const auto ssId = context.SS->SelfTabletId();
+
+        const TString& parentPathStr = Transaction.GetWorkingDir();
+        const auto& externalTableDescription = Transaction.GetCreateExternalTable();
+        const TString& name = externalTableDescription.GetName();
+
+        LOG_N("TReplaceExternalTable Propose"
+            << ": opId# " << OperationId
+            << ", path# " << parentPathStr << "/" << name << ", ReplaceIfExists:" << externalTableDescription.GetReplaceIfExists());
+
+        auto result = MakeHolder<TProposeResponse>(NKikimrScheme::StatusAccepted,
+                                                   static_cast<ui64>(OperationId.GetTxId()),
+                                                   static_cast<ui64>(ssId));
+
+        const auto parentPath = TPath::Resolve(parentPathStr, context.SS);
+        RETURN_RESULT_UNLESS(NExternalTable::IsParentPathValid(result, parentPath));
+
+        const TString acl = Transaction.GetModifyACL().GetDiffACL();
+        TPath dstPath     = parentPath.Child(name);
+        RETURN_RESULT_UNLESS(IsDestinationPathValid(result, dstPath, acl));
+
+        const auto dataSourcePath =
+            TPath::Resolve(externalTableDescription.GetDataSourcePath(), context.SS);
+        RETURN_RESULT_UNLESS(IsDataSourcePathValid(result, dataSourcePath));
+
+        const auto externalDataSource =
+            context.SS->ExternalDataSources.Value(dataSourcePath->PathId, nullptr);
+        RETURN_RESULT_UNLESS(IsDataSourceValid(result, externalDataSource));
+
+        RETURN_RESULT_UNLESS(IsApplyIfChecksPassed(result, context));
+
+        RETURN_RESULT_UNLESS(IsExternalTableDescriptionValid(result,
+                                                             externalDataSource->SourceType,
+                                                             externalTableDescription));
+
+        /// Extract old data source
+        TExternalDataSourceInfo::TPtr oldDataSource;
+        {
+            const auto oldExternalTableRecord = context.SS->ExternalTables.Value(dstPath->PathId, nullptr);
+            Y_ABORT_UNLESS(oldExternalTableRecord);
+            const auto oldDataSourcePath = TPath::Resolve(oldExternalTableRecord->DataSourcePath, context.SS);
+            RETURN_RESULT_UNLESS(IsDataSourcePathValid(result, oldDataSourcePath));
+
+            OldDataSourcePathId = oldDataSourcePath->PathId;
+            IsSameDataSource = oldDataSourcePath.PathString() == dataSourcePath.PathString();
+            if (!IsSameDataSource) {
+                oldDataSource = context.SS->ExternalDataSources.Value(oldDataSourcePath->PathId, nullptr);
+                Y_ABORT_UNLESS(oldDataSource);
```

Comment:
Is it OK that we abort process here?

Reply:
Yes. We have validated that `EXTERNAL DATA SOURCE` with such name exists using `IsDataSourcePathValid` before